google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/read_rows/StateMachine.java

```diff
@@ -0,0 +1,414 @@
+   */
+  boolean isRowInProgress() {
+    return currentState != AWAITING_NEW_ROW;
+  }
+
+  private void reset() {
+    currentState = AWAITING_NEW_ROW;
+    rowKey = null;
+    familyName = null;
+    qualifier = null;
+    timestamp = 0;
+    labels = null;
+    expectedCellSize = 0;
+    remainingCellBytes = 0;
+    completeRow = null;
+
+    adapter.reset();
+  }
+
+  /**
+   * Base class for all of the state machine's internal states.
+   *
+   * <p>Each state can consume 2 events: lastScannedRowKey and a chunk. By default, the default
+   * implementation will just throw an IllegalStateException unless the subclass adds explicit
+   * handling for the event by overriding {@link #handleLastScannedRow(ByteString)} or {@link
+   * #handleChunk(CellChunk)}.
+   */
+  abstract static class State {
+    /**
+     * Accepts the last row key scanned by the server. And set its as the last complete row. This
+     * row should be treated specially because it doesn't actually contain data.
+     *
+     * @throws IllegalStateException If the subclass can't handle last scanned row events.
+     */
+    State handleLastScannedRow(ByteString rowKey) {
+      throw new IllegalStateException();
+    }
+
+    /**
+     * Accepts a new chunk and transitions to the next state.
+     *
+     * @param chunk The new chunk to process.
+     * @return The next state.
+     * @throws IllegalStateException If the subclass can't handle chunks.
+     * @throws InvalidInputException If the subclass determines that this chunk is invalid.
+     */
+    State handleChunk(CellChunk chunk) {
+      throw new IllegalStateException();
+    }
+  }
+
+  /**
+   * The default state when the state machine is awaiting a chunk to start a new row. It will notify
+   * the adapter of the new row and delegate to the AWAITING_NEW_CELL state to process the first
+   * Cell in the Chunk. Exit states: same as AWAITING_NEW_CELL
+   */
+  private final State AWAITING_NEW_ROW =
+      new State() {
+        @Override
+        State handleLastScannedRow(ByteString rowKey) {
+          completeRow = adapter.createScanMarkerRow(rowKey);
+          lastCompleteRowKey = rowKey;
+          return AWAITING_ROW_CONSUME;
+        }
+
+        @Override
+        State handleChunk(CellChunk chunk) {
+          validate(!chunk.getResetRow(), "AWAITING_NEW_ROW: can't reset");
+          validate(!chunk.getRowKey().isEmpty(), "AWAITING_NEW_ROW: rowKey missing");
+          validate(chunk.hasFamilyName(), "AWAITING_NEW_ROW: family missing");
+          validate(chunk.hasQualifier(), "AWAITING_NEW_ROW: qualifier missing");
+          if (lastCompleteRowKey != null) {
+            validate(
+                ByteStringComparator.INSTANCE.compare(lastCompleteRowKey, chunk.getRowKey()) < 0,
+                "AWAITING_NEW_ROW: key must be strictly increasing");
+          }
+
+          rowKey = chunk.getRowKey();
+          adapter.startRow(rowKey);
+
+          // auto transition
```

Comment:
nit: sort of a vague comment, and missing leading capital.

Reply:
expanded the doc